File: apps/api/app/schemas/trip.py
"""Trip Pydantic schema — `docs/api/trips.md`."""

from __future__ import annotations

import uuid
from datetime import date as Date
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, EmailStr, Field, field_validator, model_validator

from app.schemas.poi import PoiRiseSetResponse
from app.schemas.storage import AttachmentResponse

RegionSource = Literal["manual", "poi_snapshot", "geocoded"]


class TripBase(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str | None = None
    region_hint: str | None = Field(default=None, max_length=120)
    primary_region_code: str | None = Field(
        default=None,
        min_length=2,
        max_length=10,
        pattern=r"^[0-9]{2,10}$",
    )
    start_date: Date | None = None
    end_date: Date | None = None
    visibility: Literal["private", "unlisted", "public"] = "private"


class TripCompanionInvite(BaseModel):
    email: EmailStr
    display_name: str | None = Field(default=None, max_length=80)
    role: Literal["co_owner", "editor", "viewer"] = "editor"


class TripCompanionResponse(BaseModel):
    companion_id: uuid.UUID
    trip_id: uuid.UUID
    user_id: uuid.UUID | None
    invited_email: EmailStr | None
    invited_nickname: str | None
    role: Literal["co_owner", "editor", "viewer"]
    invited_at: datetime
    joined_at: datetime | None
    created_at: datetime
    updated_at: datetime


class TripCommentCreate(BaseModel):
    body: str = Field(min_length=1, max_length=2000)
    target_type: Literal["trip", "day", "poi"] = "trip"
    target_id: uuid.UUID | None = None
    day_index: int | None = Field(default=None, ge=1)

    @field_validator("body")
    @classmethod
    def _strip_body(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("댓글 본문을 입력하세요.")
        return stripped


class TripCommentResponse(BaseModel):
    comment_id: uuid.UUID
    trip_id: uuid.UUID
    author_user_id: uuid.UUID | None
    body: str
    target_type: Literal["trip", "day", "poi"]
    target_id: uuid.UUID | None
    day_index: int | None
    created_at: datetime
    updated_at: datetime


class TripCreate(TripBase):
    companions: list[TripCompanionInvite] = Field(default_factory=list)

    @model_validator(mode="after")
    def _check_date_range(self) -> TripCreate:
        if self.start_date is None and self.end_date is None:
            return self
        if self.start_date is None or self.end_date is None:
            raise ValueError("start_date와 end_date는 동시에 채워지거나 동시에 비어야 합니다.")
        if self.end_date < self.start_date:
            raise ValueError("end_date는 start_date 이후여야 합니다.")
        return self

    @model_validator(mode="after")
    def _check_unique_companion_emails(self) -> TripCreate:
        emails = [str(companion.email).lower() for companion in self.companions]
        if len(emails) != len(set(emails)):
            raise ValueError("동반자 이메일이 중복되었습니다.")
        return self


class TripUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    region_hint: str | None = Field(default=None, max_length=120)
    primary_region_code: str | None = Field(
        default=None,
        min_length=2,
        max_length=10,
        pattern=r"^[0-9]{2,10}$",
    )
    cover_attachment_id: uuid.UUID | None = None
    start_date: Date | None = None
    end_date: Date | None = None
    visibility: Literal["private", "unlisted", "public"] | None = None
    status: Literal["draft", "planned", "in_progress", "completed", "archived"] | None = None


class TripDeleteRequest(BaseModel):
    mode: Literal["soft_delete", "transfer_leader"] = "soft_delete"
    new_owner_user_id: uuid.UUID | None = None

    @model_validator(mode="after")
    def _check_transfer_target(self) -> TripDeleteRequest:
        if self.mode == "transfer_leader" and self.new_owner_user_id is None:
            raise ValueError("transfer_leader에는 new_owner_user_id가 필요합니다.")
        return self


class TripDayCreate(BaseModel):
    day_index: int = Field(ge=1)
    date: Date | None = None
    title: str | None = Field(default=None, max_length=200)
    note: str | None = None


class TripDayUpdate(BaseModel):
    date: Date | None = None
    title: str | None = Field(default=None, max_length=200)
    note: str | None = None


class TripDayResponse(BaseModel):
    trip_id: uuid.UUID
    day_index: int
    date: Date | None
    title: str | None
    note: str | None
    created_at: datetime
    updated_at: datetime


class TripCopyRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    scope: Literal["all", "day", "range"] = "all"
    day_index: int | None = Field(default=None, ge=1)
    start_day_index: int | None = Field(default=None, ge=1)
    end_day_index: int | None = Field(default=None, ge=1)
    date_shift_days: int = 0
    target_trip_id: uuid.UUID | None = None

    @model_validator(mode="after")
    def _check_scope(self) -> TripCopyRequest:
        if self.scope == "day" and self.day_index is None:
            raise ValueError("scope=day에는 day_index가 필요합니다.")
        if self.scope == "range":
            if self.start_day_index is None or self.end_day_index is None:
                raise ValueError("scope=range에는 start_day_index/end_day_index가 필요합니다.")
            if self.end_day_index < self.start_day_index:
                raise ValueError("end_day_index는 start_day_index 이후여야 합니다.")
        return self


class TripResponse(BaseModel):
    trip_id: uuid.UUID
    owner_user_id: uuid.UUID
    title: str
    description: str | None
    region_hint: str | None
    primary_region_code: str | None
    primary_region_source: RegionSource | None
    start_date: Date | None
    end_date: Date | None
    visibility: Literal["private", "unlisted", "public"]
    status: Literal["draft", "planned", "in_progress", "completed", "archived"]
    version: int
    created_at: datetime
    updated_at: datetime


class TripCopyResponse(BaseModel):
    trip: TripResponse
    created_trip: bool
    copied_day_count: int
    copied_poi_count: int
    copied_attachment_count: int


class TripViewPoi(BaseModel):
    poi_id: uuid.UUID
    feature_id: str | None
    sort_order: str
    title: str | None
    feature: dict[str, Any]
    marker_color: str | None
    marker_icon: str | None
    is_broken: bool
    user_note: str | None
    planned_arrival_at: datetime | None
    planned_departure_at: datetime | None
    budget_amount: Decimal | None
    actual_amount: Decimal | None
    currency: str
    user_url: str | None
    rise_set: PoiRiseSetResponse | None
    feature_link_broken_at: datetime | None
    version: int
    created_at: datetime
    updated_at: datetime


class TripViewDay(BaseModel):
    day_index: int
    date: Date | None
    title: str | None
    pois: list[TripViewPoi]


class TripViewShareLink(BaseModel):
    share_id: uuid.UUID
    visibility: Literal["view_only", "comment", "edit"]
    expires_at: datetime | None
    revoked_at: datetime | None
    last_used_at: datetime | None
    created_at: datetime


class TripView(BaseModel):
    trip: TripResponse
    days: list[TripViewDay]
    companions: list[TripCompanionResponse]
    share_links: list[TripViewShareLink]
    broken_feature_count: int


class TripSharedView(BaseModel):
    visibility: Literal["view_only", "comment", "edit"]
    trip: TripResponse
    days: list[TripViewDay]
    broken_feature_count: int


class TripDayOptimizeRequest(BaseModel):
    strategy: Literal["nearest_neighbor"] = "nearest_neighbor"
    start_poi_id: uuid.UUID | None = None
    persist: bool = False


class TripDayOptimizeMove(BaseModel):
    poi_id: uuid.UUID
    old_sort_order: str
    new_sort_order: str


class TripDayOptimizeResponse(BaseModel):
    trip_id: uuid.UUID
    day_index: int
    ordered_poi_ids: list[uuid.UUID]
    moves: list[TripDayOptimizeMove]
    distance_meters: int | None
    warnings: list[str] = Field(default_factory=list)


class TripDistanceMatrixResponse(BaseModel):
    trip_id: uuid.UUID
    day_index: int
    poi_ids: list[uuid.UUID]
    distances_meters: list[list[int | None]]
    warnings: list[str] = Field(default_factory=list)


class TripAttachmentResponse(AttachmentResponse):
    pass

File: packages/schemas/src/trip.ts
import { z } from 'zod';
import { Iso8601Schema, NonNegativeDecimalStringSchema } from './common';
import { PoiRiseSetResponseSchema } from './poi';
import { AttachmentCreateSchema, AttachmentResponseSchema } from './storage';

/** `docs/api/trips.md`. */
export const TripStatusSchema = z.enum([
  'draft',
  'planned',
  'in_progress',
  'completed',
  'archived',
]);
export const TripVisibilitySchema = z.enum(['private', 'unlisted', 'public']);
const RegionCodeSchema = z.string().regex(/^[0-9]{2,10}$/);
export const TripPrimaryRegionSourceSchema = z.enum(['manual', 'poi_snapshot', 'geocoded']);
export const TripCompanionRoleSchema = z.enum(['co_owner', 'editor', 'viewer']);
export const TripShareLinkVisibilitySchema = z.enum(['view_only', 'comment', 'edit']);
export type TripStatus = z.infer<typeof TripStatusSchema>;
export type TripVisibility = z.infer<typeof TripVisibilitySchema>;
export type TripPrimaryRegionSource = z.infer<typeof TripPrimaryRegionSourceSchema>;
export type TripCompanionRole = z.infer<typeof TripCompanionRoleSchema>;
export type TripShareLinkVisibility = z.infer<typeof TripShareLinkVisibilitySchema>;

export const TripCompanionInviteSchema = z.object({
  email: z.string().email().max(320),
  display_name: z.string().max(80).nullable().optional(),
  role: TripCompanionRoleSchema.default('editor'),
});
export type TripCompanionInvite = z.infer<typeof TripCompanionInviteSchema>;

export const TripCompanionResponseSchema = z.object({
  companion_id: z.string().uuid(),
  trip_id: z.string().uuid(),
  user_id: z.string().uuid().nullable(),
  invited_email: z.string().email().nullable(),
  invited_nickname: z.string().nullable(),
  role: TripCompanionRoleSchema,
  invited_at: Iso8601Schema,
  joined_at: Iso8601Schema.nullable(),
  created_at: Iso8601Schema,
  updated_at: Iso8601Schema,
});
export type TripCompanionResponse = z.infer<typeof TripCompanionResponseSchema>;

export const TripCommentTargetSchema = z.enum(['trip', 'day', 'poi']);
export const TripCommentCreateSchema = z.object({
  body: z.string().trim().min(1).max(2000),
  target_type: TripCommentTargetSchema.default('trip'),
  target_id: z.string().uuid().nullable().optional(),
  day_index: z.number().int().min(1).nullable().optional(),
});
export type TripCommentCreate = z.infer<typeof TripCommentCreateSchema>;

export const TripCommentResponseSchema = z.object({
  comment_id: z.string().uuid(),
  trip_id: z.string().uuid(),
  author_user_id: z.string().uuid().nullable(),
  body: z.string(),
  target_type: TripCommentTargetSchema,
  target_id: z.string().uuid().nullable(),
  day_index: z.number().int().nullable(),
  created_at: Iso8601Schema,
  updated_at: Iso8601Schema,
});
export type TripCommentResponse = z.infer<typeof TripCommentResponseSchema>;

export const TripShareLinkCreateSchema = z.object({
  visibility: TripShareLinkVisibilitySchema.default('view_only'),
  expires_at: Iso8601Schema.nullable().optional(),
});
export type TripShareLinkCreate = z.infer<typeof TripShareLinkCreateSchema>;

export const TripShareLinkResponseSchema = z.object({
  share_id: z.string().uuid(),
  trip_id: z.string().uuid(),
  visibility: TripShareLinkVisibilitySchema,
  token: z.string(),
  url: z.string().url(),
  expires_at: Iso8601Schema.nullable(),
  revoked_at: Iso8601Schema.nullable(),
  last_used_at: Iso8601Schema.nullable(),
  created_at: Iso8601Schema,
});
export type TripShareLinkResponse = z.infer<typeof TripShareLinkResponseSchema>;

export const TripCreateSchema = z.object({
  title: z.string().min(1).max(200),
  description: z.string().nullable().optional(),
  region_hint: z.string().max(120).nullable().optional(),
  primary_region_code: RegionCodeSchema.nullable().optional(),
  start_date: z.string().date().nullable().optional(),
  end_date: z.string().date().nullable().optional(),
  visibility: TripVisibilitySchema.default('private'),
  companions: z.array(TripCompanionInviteSchema).default([]),
});
export type TripCreate = z.infer<typeof TripCreateSchema>;

export const TripUpdateSchema = z.object({
  title: z.string().min(1).max(200).optional(),
  description: z.string().nullable().optional(),
  region_hint: z.string().max(120).nullable().optional(),
  primary_region_code: RegionCodeSchema.nullable().optional(),
  cover_attachment_id: z.string().uuid().nullable().optional(),
  start_date: z.string().date().nullable().optional(),
  end_date: z.string().date().nullable().optional(),
  visibility: TripVisibilitySchema.optional(),
  status: TripStatusSchema.optional(),
});
export type TripUpdate = z.infer<typeof TripUpdateSchema>;

export const TripDeleteRequestSchema = z
  .object({
    mode: z.enum(['soft_delete', 'transfer_leader']).default('soft_delete'),
    new_owner_user_id: z.string().uuid().nullable().optional(),
  })
  .superRefine((value, ctx) => {
    if (value.mode === 'transfer_leader' && !value.new_owner_user_id) {
      ctx.addIssue({
        code: 'custom',
        path: ['new_owner_user_id'],
        message: 'new_owner_user_id is required for transfer_leader',
      });
    }
  });
export type TripDeleteRequest = z.infer<typeof TripDeleteRequestSchema>;

export const TripResponseSchema = z.object({
  trip_id: z.string().uuid(),
  owner_user_id: z.string().uuid(),
  title: z.string(),
  description: z.string().nullable(),
  region_hint: z.string().nullable(),
  primary_region_code: RegionCodeSchema.nullable(),
  primary_region_source: TripPrimaryRegionSourceSchema.nullable(),
  start_date: z.string().date().nullable(),
  end_date: z.string().date().nullable(),
  visibility: TripVisibilitySchema,
  status: TripStatusSchema,
  version: z.number().int(),
  created_at: Iso8601Schema,
  updated_at: Iso8601Schema,
});
export type TripResponse = z.infer<typeof TripResponseSchema>;

// 일자/POI 마커 색 override — 팔레트 키 P-01~P-16 (ADR-055).
const MarkerColorKeySchema = z.string().regex(/^P-(0[1-9]|1[0-6])$/, 'marker color는 P-01~P-16 형식.');

export const TripDayCreateSchema = z.object({
  day_index: z.number().int().min(1).optional(),
  date: z.string().date().nullable().optional(),
  title: z.string().max(200).nullable().optional(),
  note: z.string().nullable().optional(),
  marker_color: MarkerColorKeySchema.nullable().optional(),
});
export type TripDayCreate = z.infer<typeof TripDayCreateSchema>;

export const TripDayUpdateSchema = z.object({
  date: z.string().date().nullable().optional(),
  title: z.string().max(200).nullable().optional(),
  note: z.string().nullable().optional(),
  marker_color: MarkerColorKeySchema.nullable().optional(),
});
export type TripDayUpdate = z.infer<typeof TripDayUpdateSchema>;

export const TripDayResponseSchema = z.object({
  trip_id: z.string().uuid(),
  day_index: z.number().int(),
  date: z.string().date().nullable(),
  title: z.string().nullable(),
  note: z.string().nullable(),
  // 일자 색 override(팔레트 키). NULL이면 인덱스 기본색으로 파생(ADR-055).
  marker_color: z.string().nullable().default(null),
  // backend는 항상 version을 보낸다. default는 version 컬럼 도입(T-287) 이전 응답/목업 호환용.
  version: z.number().int().default(1),
  created_at: Iso8601Schema,
  updated_at: Iso8601Schema,
});
export type TripDayResponse = z.infer<typeof TripDayResponseSchema>;

export const TripCopyRequestSchema = z
  .object({
    title: z.string().min(1).max(200).nullable().optional(),
    scope: z.enum(['all', 'day', 'range']).default('all'),
    day_index: z.number().int().min(1).nullable().optional(),
    start_day_index: z.number().int().min(1).nullable().optional(),
    end_day_index: z.number().int().min(1).nullable().optional(),
    date_shift_days: z.number().int().default(0),
    target_trip_id: z.string().uuid().nullable().optional(),
  })
  .superRefine((value, ctx) => {
    if (value.scope === 'day' && !value.day_index) {
      ctx.addIssue({ code: 'custom', path: ['day_index'], message: 'day_index is required' });
    }
    if (value.scope === 'range') {
      if (!value.start_day_index || !value.end_day_index) {
        ctx.addIssue({
          code: 'custom',
          path: ['start_day_index'],
          message: 'start_day_index/end_day_index are required',
        });
      } else if (value.end_day_index < value.start_day_index) {
        ctx.addIssue({
          code: 'custom',
          path: ['end_day_index'],
          message: 'end_day_index must be greater than or equal to start_day_index',
        });
      }
    }
  });
export type TripCopyRequest = z.infer<typeof TripCopyRequestSchema>;

export const TripCopyResponseSchema = z.object({
  trip: TripResponseSchema,
  created_trip: z.boolean(),
  copied_day_count: z.number().int().nonnegative(),
  copied_poi_count: z.number().int().nonnegative(),
  copied_attachment_count: z.number().int().nonnegative(),
});
export type TripCopyResponse = z.infer<typeof TripCopyResponseSchema>;

export const TripViewPoiSchema = z.object({
  poi_id: z.string().uuid(),
  feature_id: z.string().nullable(),
  sort_order: z.string(),
  title: z.string().nullable(),
  feature: z.record(z.string(), z.unknown()),
  marker_color: z.string().nullable(),
  marker_icon: z.string().nullable(),
  // 지도 핀·목록 뱃지 parity용 서버 계산 색(custom > 일자색). 항상 유효 팔레트 키(ADR-055).
  display_marker_color: z.string().nullable().default(null),
  is_broken: z.boolean(),
  user_note: z.string().nullable(),
  planned_arrival_at: Iso8601Schema.nullable(),
  planned_departure_at: Iso8601Schema.nullable(),
  budget_amount: NonNegativeDecimalStringSchema.nullable(),
  actual_amount: NonNegativeDecimalStringSchema.nullable(),
  currency: z.string().regex(/^[A-Z]{3}$/),
  user_url: z.string().nullable(),
  rise_set: PoiRiseSetResponseSchema.nullable(),
  feature_link_broken_at: Iso8601Schema.nullable(),
  version: z.number().int(),
  created_at: Iso8601Schema,
  updated_at: Iso8601Schema,
});
export type TripViewPoi = z.infer<typeof TripViewPoiSchema>;

export const TripDayHolidaySchema = z.object({
  date: z.string().date(),
  name: z.string(),
  dataset: z.enum([
    'holidays',
    'national_holidays',
    'anniversaries',
    'solar_terms_24',
    'sundry_days',
  ]),
});
export type TripDayHoliday = z.infer<typeof TripDayHolidaySchema>;

export const TripViewDaySchema = z.object({
  day_index: z.number().int(),
  date: z.string().date().nullable(),
  // date는 override-only, effective_date는 파생(override 또는 start_date+day_index) — ADR-055.
  effective_date: z.string().date().nullable().default(null),
  // effective_date가 여행 [start,end] 밖이면 true(F1: 기간 축소 시 경고).
  out_of_range: z.boolean().default(false),
  // 일자 색 override(팔레트 키). NULL이면 인덱스 기본색으로 파생.
  marker_color: z.string().nullable().default(null),
  title: z.string().nullable(),
  // backend는 항상 version을 보낸다. default는 version 컬럼 도입(T-287) 이전 응답/목업 호환용.
  version: z.number().int().default(1),
  holidays: z.array(TripDayHolidaySchema).default([]),
  pois: z.array(TripViewPoiSchema),
});
export type TripViewDay = z.infer<typeof TripViewDaySchema>;

export const TripViewShareLinkSchema = z.object({
  share_id: z.string().uuid(),
  visibility: TripShareLinkVisibilitySchema,
  expires_at: Iso8601Schema.nullable(),
  revoked_at: Iso8601Schema.nullable(),
  last_used_at: Iso8601Schema.nullable(),
  created_at: Iso8601Schema,
});
export type TripViewShareLink = z.infer<typeof TripViewShareLinkSchema>;

export const TripViewSchema = z.object({
  trip: TripResponseSchema,
  days: z.array(TripViewDaySchema),
  companions: z.array(TripCompanionResponseSchema),
  share_links: z.array(TripViewShareLinkSchema),
  broken_feature_count: z.number().int().nonnegative(),
});
export type TripView = z.infer<typeof TripViewSchema>;

export const TripSharedViewSchema = z.object({
  visibility: TripShareLinkVisibilitySchema,
  trip: TripResponseSchema,
  days: z.array(TripViewDaySchema),
  broken_feature_count: z.number().int().nonnegative(),
});
export type TripSharedView = z.infer<typeof TripSharedViewSchema>;

export const TripDayOptimizeRequestSchema = z.object({
  // two_opt = nearest-neighbor seed + 2-opt local search (스마트 정렬, 기본).
  strategy: z.enum(['nearest_neighbor', 'two_opt']).default('two_opt'),
  start_poi_id: z.string().uuid().nullable().optional(),
  persist: z.boolean().default(false),
});
export type TripDayOptimizeRequest = z.infer<typeof TripDayOptimizeRequestSchema>;

export const TripDayOptimizeMoveSchema = z.object({
  poi_id: z.string().uuid(),
  old_sort_order: z.string(),
  new_sort_order: z.string(),
});
export type TripDayOptimizeMove = z.infer<typeof TripDayOptimizeMoveSchema>;

export const TripDayOptimizeResponseSchema = z.object({
  trip_id: z.string().uuid(),
  day_index: z.number().int(),
  ordered_poi_ids: z.array(z.string().uuid()),
  moves: z.array(TripDayOptimizeMoveSchema),
  distance_meters: z.number().int().nullable(),
  previous_distance_meters: z.number().int().nullable().default(null),
  warnings: z.array(z.string()),
});
export type TripDayOptimizeResponse = z.infer<typeof TripDayOptimizeResponseSchema>;

export const TripDistanceMatrixResponseSchema = z.object({
  trip_id: z.string().uuid(),
  day_index: z.number().int(),
  poi_ids: z.array(z.string().uuid()),
  distances_meters: z.array(z.array(z.number().int().nullable())),
  warnings: z.array(z.string()),
});
export type TripDistanceMatrixResponse = z.infer<typeof TripDistanceMatrixResponseSchema>;

export const TripAttachmentCreateSchema = AttachmentCreateSchema;
export const TripAttachmentResponseSchema = AttachmentResponseSchema;
export type TripAttachmentCreate = z.infer<typeof TripAttachmentCreateSchema>;
export type TripAttachmentResponse = z.infer<typeof TripAttachmentResponseSchema>;
